os/arch/arm/src/amebad/amebad_i2s.c

```diff
@@ -0,0 +1,1818 @@
+	 * expected 'failure' (meaning that the wait for the semaphore was
+	 * interrupted by a signal.
+	 */
+	do {
+		ret = sem_wait(&priv->bufsem_tx);
+		DEBUGASSERT(ret == 0 || errno == EINTR);
+	} while (ret < 0);
+}
+
+/****************************************************************************
+* Name: i2s_buf_tx_allocate
+*
+* Description:
+*   Allocate a buffer container by removing the one at the head of the
+*   free list
+*
+* Input Parameters:
+*   priv - i2s state instance
+*
+* Returned Value:
+*   A non-NULL pointer to the allocate buffer container on success; NULL if
+*   there are no available buffer containers.
+*
+* Assumptions:
+*   The caller does NOT have exclusive access to the I2S state structure.
+*   That would result in a deadlock!
+*
+****************************************************************************/
+
+static struct amebad_buffer_s *i2s_buf_tx_allocate(struct amebad_i2s_s *priv)
+{
+	struct amebad_buffer_s *bfcontainer;
+	irqstate_t flags;
+
+	/* Set aside a buffer container.  By doing this, we guarantee that we will
+	 * have at least one free buffer container.
+	 */
+	i2s_bufsem_tx_take(priv);
+
+	/* Get the buffer from the head of the free list */
+	flags = irqsave();
+	bfcontainer = priv->freelist_tx;
+	ASSERT(bfcontainer);
+
+	/* Unlink the buffer from the freelist */
+	priv->freelist_tx = bfcontainer->flink;
+	irqrestore(flags);
+
+	return bfcontainer;
+}
+
+/****************************************************************************
+* Name: i2s_buf_tx_free
+*
+* Description:
+*   Free buffer container by adding it to the head of the free list
+*
+* Input Parameters:
+*   priv - I2S state instance
+*   bfcontainer - The buffer container to be freed
+*
+* Returned Value:
+*   None
+*
+* Assumptions:
+*   The caller has exclusive access to the I2S state structure
+*
+****************************************************************************/
+
+static void i2s_buf_tx_free(struct amebad_i2s_s *priv, struct amebad_buffer_s *bfcontainer)
+{
+	irqstate_t flags;
+
+	/* Put the buffer container back on the free list */
+	flags = irqsave();
+	bfcontainer->flink = priv->freelist_tx;
+	priv->freelist_tx = bfcontainer;
+	irqrestore(flags);
+
+	/* Wake up any threads waiting for a buffer container */
+	i2s_bufsem_tx_give(priv);
+}
+
+/****************************************************************************
+ * Name: i2s_buf_tx_initialize
+ *
+ * Description:
+ *   Initialize the buffer container allocator by adding all of the
+ *   pre-allocated buffer containers to the free list
+ *
+ * Input Parameters:
+ *   priv - I2S state instance
+ *
+ * Returned Value:
+ *   None
+ *
+ * Assumptions:
+ *   Called early in I2S initialization so that there are no issues with
+ *   concurrency.
+ *
+ ****************************************************************************/
+
+static void i2s_buf_tx_initialize(struct amebad_i2s_s *priv)
+{
+	int i;
+
+	priv->freelist_tx = NULL;
```

Comment:
Is it okay not to check that `priv` is NULL or not?

Reply:
The `i2s_buf_tx_initialize` has been called by `amebad_i2s_initialize` which also checks for !priv before calling this function.